DOCSP-13838 crud projection (#66)
mongodb/docs-golang: source/fundamentals/crud/read-operations/project.txt

==============================
Specify Which Fields to Return
==============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> In this guide, you can learn how to specify whether fields in a
> document should return from read operations.

Reviewer: Consider shifting the tone a bit to something the user is controlling.

```suggestion
In this guide, you can learn how to specify which fields in a
document are returned by read operations.
```

Sample Data
~~~~~~~~~~~

> Run the following snippet to load the documents into the ``ratings``
> collection of the ``tea`` database:

Reviewer: Consider stating the goal first, and then how to achieve it.

Something like:

```suggestion
To load documents into the ``ratings`` collection of the ``tea`` database,
run the following snippet:
```

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/projection.go 
   :language: go
   :dedent:
   :start-after: begin insertDocs
   :end-before: end insertDocs

.. include:: /includes/fundamentals/tea-sample-data-ending.rst

> Project

Reviewer: [question]

This is the first time we're mentioning the idea of "projection" on this page. Up until now, we've been saying "specify which fields to return". Do you think we need a sentence explaining that projection _is_ specifying which fields to return?

If I'm a user who has no idea what projection is, will I be confused without seeing a definition of projection? I'm not sure this is 100% necessary, but want to raise it for consideration.

-------

To specify whether fields in a document should return, pass an interface

> specifying the field and ``0`` or ``1`` to the ``SetProjection()``

Reviewer: Do you think we need an explanation that ``0`` means the field is not included and ``1`` means that is included? I see that this information is included in the sections further down "Exclude a Field" and "Include a Field", but I'm wondering if we should move it further up to here where we kind of introduce the syntax. As currently written, when I read this section, I'm wondering what the `0` and `1` correspond to.

   [{_id ObjectID("...")} {type Assam}]
   [{_id ObjectID("...")} {type Oolong}]
   [{_id ObjectID("...")} {type Earl Grey}]
   [{_id ObjectID("...")} {type English Breakfast}]

Include a Field
~~~~~~~~~~~~~~~

Specify a ``1`` in the interface passed to the ``SetProjection()``
function to include a field. For all fields you don't explicitly list
in the interface, the driver excludes them.

.. important::

    You can exclude the ``_id`` field even if you specified to include
    certain fields.

.. _go-include-projection:

Example
```````

The following example performs the following projection on the matched
documents from the ``Find()`` function:

- Include the ``type`` and ``rating`` field
- Exclude the ``_id`` field

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/projection.go
   :language: go
   :dedent:
   :start-after: begin include projection
   :end-before: end include projection

After running this example, the output resembles the following:

.. code-block:: none
   :copyable: false

   [{type Masala} {rating 10}]
   [{type Assam} {rating 5}]
   [{type Oolong} {rating 7}]
   [{type Earl Grey} {rating 8}]
   [{type English Breakfast} {rating 5}]

> .. tip:: Using Aggregation 

Reviewer: I would consider having this example stand on its own rather than in a tip box. I think it's a good example to include and should be treated the same as the others.

   You can also include the :manual:`$project </reference/operator/aggregation/project/>`
   stage to specify a projection in an aggregation pipeline.

>    This example specifies the same projection from the preceding example

Reviewer: Rather than having users have to scroll back to the previous example, consider reiterating what the projection does. You repeat text, but it will help users who scroll quickly through and stop on this example to not have them scrolling back and forth.

==============================
Specify Which Fields to Return
==============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to specify which fields to return in a
document from read operations.

Sample Data
~~~~~~~~~~~

To run the examples in this guide, load these documents into the
``ratings`` collection of the ``tea`` database with the following
snippet:

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/projection.go 
   :language: go
   :dedent:
   :start-after: begin insertDocs
   :end-before: end insertDocs

.. include:: /includes/fundamentals/tea-sample-data-ending.rst

Projection
----------

A projection specifies which fields to return in a document. To create a
projection, pass an interface specifying the field and whether to
indclude that field to the ``SetProjection()`` function of a read
operation's options. To include the field, specify a ``1`` and to

> exclude a field, specify a ``0``. 

Reviewer 2: This paragraph is kinda difficult to grok. The description of the structure of the projection document and usage in the Go driver API are much better illustrated in the examples below and can be excluded here.

Example:
> A [projection](https://docs.mongodb.com/manual/reference/glossary/#std-term-projection) is a document that specifies which fields to return in query results. A projection document includes field names followed by a ``1`` (to include) or ``0`` (to exclude). See the examples below for more details on the projection document format and how to use projections in the Go driver.

Specify the options as the last parameter to the following read
operation functions:

- ``Find()``
- ``FindOne()``
- ``FindOneAndDelete()``
- ``FindOneAndReplace()``

> - ``FindOneAndUpdate()``

Reviewer 2: Should this list include `Aggregate()` as well?
